src/rust/engine/fs/src/lib.rs

```diff
@@ -714,12 +837,13 @@ pub trait VFS<E: Send + Sync + 'static>: Clone + Send + Sync + 'static {
       return future::ok(vec![]).to_boxed();
     }
 
```

Comment:
Darn, sorry... this was more challenging than I thought it would be. I'm worried that the result is too complicated right now (because the post expansion loop essentially ends up recreating/rexecuting the expansion logic, the same way you were recreating it in the python code).

In theory tracking associations all the way through `(input glob string -> PathGlob -> PathStat)` should allow us to avoid reconstructing the match logic afterward. It's also possible I'm missing something: would be happy to pair on this to try and figure it out together.

I also expect that this would be easier with (memoized) recursion, and wonder whether we shouldn't land a separate PR to switch this to memoized recursion before tackling this.

Reply:
In the previous `warn-globs-v2` branch I messaged in the slack, I had done exactly that, by having a Vec of "source" PathGlobs stored in an IndexMap (so we can iterate in reverse to do this in topological order), which would memoize exactly what we describe. I can reintroduce that now that the other kinks are worked out.

Reply:
I can also land that in a separate PR, though.

Reply:
This should be fixed now -- we merely record the "source" as a `GlobSource` instance, for each glob in `completed`, which can be either another `PathGlob`, or a `GlobParsedSource`, and iterate in reverse over `completed`.